lock-api-objects/src/main/java/com/palantir/lock/watch/LockMetadata.java

```diff
@@ -0,0 +1,142 @@
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.palantir.lock.watch.LockMetadata.Created;
+import com.palantir.lock.watch.LockMetadata.Deleted;
+import com.palantir.lock.watch.LockMetadata.Unchanged;
+import com.palantir.lock.watch.LockMetadata.Updated;
+import org.immutables.value.Value;
+import org.immutables.value.Value.Immutable;
+
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+@JsonSubTypes({
+    @JsonSubTypes.Type(value = Unchanged.class, name = Unchanged.TYPE),
+    @JsonSubTypes.Type(value = Updated.class, name = Updated.TYPE),
+    @JsonSubTypes.Type(value = Deleted.class, name = Deleted.TYPE),
+    @JsonSubTypes.Type(value = Created.class, name = Created.TYPE),
+})
+public interface LockMetadata {
+    <T> T accept(Visitor<T> visitor);
+
+    static Unchanged unchanged() {
+        return ImmutableUnchanged.builder().build();
+    }
+
+    static Updated update(byte[] oldValue, byte[] newValue) {
+        return ImmutableUpdated.of(oldValue, newValue);
+    }
+
+    static Deleted delete(byte[] deletedValue) {
+        return ImmutableDeleted.of(deletedValue);
+    }
+
+    static Created create(byte[] createdValue) {
+        return ImmutableCreated.of(createdValue);
+    }
+
+    interface Visitor<T> {
+        T visit(Unchanged unchanged);
+
+        T visit(Updated updated);
+
+        T visit(Deleted deleted);
+
+        T visit(Created created);
+    }
+
+    /**
+     * LockMetadata event indicating that a value has definitely not been changed.
+     */
+    @Immutable
+    @JsonSerialize(as = ImmutableUnchanged.class)
+    @JsonDeserialize(as = ImmutableUnchanged.class)
+    @JsonTypeName(Unchanged.TYPE)
+    interface Unchanged extends LockMetadata {
+        String TYPE = "unchanged";
+
+        @Override
+        default <T> T accept(Visitor<T> visitor) {
+            return visitor.visit(this);
+        }
+    }
+
+    /**
+     * LockMetadata event indicating that an existing value might have changed.
+     */
+    @Immutable
+    @JsonSerialize(as = ImmutableUpdated.class)
+    @JsonDeserialize(as = ImmutableUpdated.class)
+    @JsonTypeName(Updated.TYPE)
+    interface Updated extends LockMetadata {
+        String TYPE = "update";
+
+        @Value.Parameter
+        byte[] oldValue();
+
+        @Value.Parameter
+        byte[] newValue();
+
+        @Override
+        default <T> T accept(Visitor<T> visitor) {
+            return visitor.visit(this);
+        }
+    }
+
+    /**
+     * LockMetadata event indicating that a value might have been deleted (it might no longer exists).
+     */
+    @Immutable
+    @JsonSerialize(as = ImmutableDeleted.class)
+    @JsonDeserialize(as = ImmutableDeleted.class)
+    @JsonTypeName(Deleted.TYPE)
+    interface Deleted extends LockMetadata {
+        String TYPE = "delete";
+
+        @Value.Parameter
+        byte[] deletedValue();
```

Comment:
Considering that the class is already called `Deleted`, `deletedValue` seems a bit redundant. Thoughts on just `value`?

Reply:
I was initially worried that just calling it `value` in this and the `Created` case might cause confusion as the value for these objects reflects a different point in time (before the operation vs. after the operation). We can retain this information, but avoid repeating the classname  by renaming to `oldValue` and `newValue` (aligns with the `Updated` case as well and is in line with the S_K and S_K^` notation introduced in the theory section of the RFC).